### PR TITLE
Redact http header values in logs

### PR DESCRIPTION
**Pull request by sfc-gh-elinardi**
Introducing redaction to rpc authz evaluation logs.

This leverages protobuf's `debug_redact` annotation. In the protobuf definition, fields which need redaction are annotated with `debug_redact`. Then, before logging the protobuf message, we look for the annotation and replace the value with `--REDACTED--`

In this PR I'm only adding the redaction to httpoverrpc header values.

Testing
```
{"logger":"sansshell-proxy","ts":"2023-10-09 14:17:51.528922","caller":{"file":"rpcauth.go","line":215},"level":2,"msg":"evaluating authz policy","method":"/Proxy.Proxy/Proxy","peer":{"net":{"network":"tcp","address":"127.0.0.1","port":"64348"},"cert":{"subject":"CN=test.example.com","issuer":"CN=example.com","dnsnames":["test.example.com"],"spiffeid":""},"principal":null},"input":{"method":"/HTTPOverRPC.HTTPOverRPC/Host","message":{"request":{"method":"POST", "request_uri":"/session", "headers":[{"key":"Content-Type", "values":["--REDACTED--"]}, {"key":"Accept", "values":["--REDACTED--"]}, {"key":"User-Agent", "values":["--REDACTED--"]}, {"key":"Authorization", "values":["--REDACTED--"]}]}, "port":8082, "hostname":"localhost", "protocol":"http"},"type":"HTTPOverRPC.HostHTTPRequest","metadata":{":authority":["test.example.com"],"content-type":["application/grpc"],"user-agent":["grpc-go/1.52.3"],"grpc-accept-encoding":["gzip"]},"peer":{"net":{"network":"tcp","address":"127.0.0.1","port":"64348"},"cert":{"subject":"CN=test.example.com","issuer":"CN=example.com","dnsnames":["test.example.com"],"spiffeid":""},"principal":null},"host":{"net":{"network":"tcp","address":"127.0.0.1","port":"50042"},"cert":{"subject":"CN=test.example.com","issuer":"CN=example.com","dnsnames":["test.example.com"],"spiffeid":""},"principal":null},"environment":{"non_host_policy_check":true},"extensions":null}}
{"logger":"sansshell-proxy","ts":"2023-10-09 14:17:51.529090","caller":{"file":"rpcauth.go","line":237},"level":2,"msg":"evaluating authz policy post hooks","method":"/Proxy.Proxy/Proxy","peer":{"net":{"network":"tcp","address":"127.0.0.1","port":"64348"},"cert":{"subject":"CN=test.example.com","issuer":"CN=example.com","dnsnames":["test.example.com"],"spiffeid":""},"principal":null},"input":{"method":"/HTTPOverRPC.HTTPOverRPC/Host","message":{"request":{"method":"POST", "request_uri":"/session", "headers":[{"key":"Content-Type", "values":["--REDACTED--"]}, {"key":"Accept", "values":["--REDACTED--"]}, {"key":"User-Agent", "values":["--REDACTED--"]}, {"key":"Authorization", "values":["--REDACTED--"]}]}, "port":8082, "hostname":"localhost", "protocol":"http"},"type":"HTTPOverRPC.HostHTTPRequest","metadata":{"user-agent":["grpc-go/1.52.3"],"grpc-accept-encoding":["gzip"],":authority":["test.example.com"],"content-type":["application/grpc"]},"peer":{"net":{"network":"tcp","address":"127.0.0.1","port":"64348"},"cert":{"subject":"CN=test.example.com","issuer":"CN=example.com","dnsnames":["test.example.com"],"spiffeid":""},"principal":null},"host":{"net":{"network":"tcp","address":"127.0.0.1","port":"50042"},"cert":{"subject":"CN=test.example.com","issuer":"CN=example.com","dnsnames":["test.example.com"],"spiffeid":""},"principal":null},"environment":{"non_host_policy_check":true},"extensions":null}}
{"logger":"sansshell-proxy","ts":"2023-10-09 14:17:51.529295","caller":{"file":"rpcauth.go","line":255},"level":0,"msg":"authz policy evaluation result","method":"/Proxy.Proxy/Proxy","peer":{"net":{"network":"tcp","address":"127.0.0.1","port":"64348"},"cert":{"subject":"CN=test.example.com","issuer":"CN=example.com","dnsnames":["test.example.com"],"spiffeid":""},"principal":null},"authorizationResult":true,"input":{"method":"/HTTPOverRPC.HTTPOverRPC/Host","message":{"request":{"method":"POST", "request_uri":"/session", "headers":[{"key":"Content-Type", "values":["--REDACTED--"]}, {"key":"Accept", "values":["--REDACTED--"]}, {"key":"User-Agent", "values":["--REDACTED--"]}, {"key":"Authorization", "values":["--REDACTED--"]}]}, "port":8082, "hostname":"localhost", "protocol":"http"},"type":"HTTPOverRPC.HostHTTPRequest","metadata":{"user-agent":["grpc-go/1.52.3"],"grpc-accept-encoding":["gzip"],":authority":["test.example.com"],"content-type":["application/grpc"]},"peer":{"net":{"network":"tcp","address":"127.0.0.1","port":"64348"},"cert":{"subject":"CN=test.example.com","issuer":"CN=example.com","dnsnames":["test.example.com"],"spiffeid":""},"principal":null},"host":{"net":{"network":"tcp","address":"127.0.0.1","port":"50042"},"cert":{"subject":"CN=test.example.com","issuer":"CN=example.com","dnsnames":["test.example.com"],"spiffeid":""},"principal":null},"environment":{"non_host_policy_check":true},"extensions":null},"denialHints":[]}
```

### CODE REVIEW

File: auth/opa/rpcauth/rpcauth.go [hunk 1]
```diff
@@ -88,6 +94,110 @@ func NewWithPolicy(ctx context.Context, policy string, authzHooks ...RPCAuthzHoo
...
+	redactedInput := RPCAuthInput{
+		Method:      input.Method,
+		MessageType: input.MessageType,
+		Metadata:    input.Metadata,
+		Peer:        input.Peer,
+		Host:        input.Host,
+		Environment: input.Environment,
+		Extensions:  input.Extensions,
+	}
+	if input.MessageType == mockMessageType || input.MessageType == "" {
```
Comment (sfc-gh-srhodes) [inline auth/opa/rpcauth/rpcauth.go:177]: Why do we need the special case for Mock?

If it's for getting around the tests passing in fake proto types, I'd rather fix that by making them pass around real proto types. You can use google.protobuf.Empty for cases where you want to leave input.Message blank.

File: auth/opa/rpcauth/rpcauth.go [hunk 2]
```diff
@@ -88,6 +94,110 @@ func NewWithPolicy(ctx context.Context, policy string, authzHooks ...RPCAuthzHoo
 	return New(p, authzHooks...), nil
 }
 
+func isMessage(descriptor protoreflect.FieldDescriptor) bool {
```
Comment (sfc-gh-srhodes) [inline auth/opa/rpcauth/rpcauth.go:97]: Organizational nit: Maybe move this to a separate file? It's mostly logically separate from the rest of rpcauth.go.

**Review by sfc-gh-srhodes — CHANGES_REQUESTED** [review]
Looks okay overall. Could you add tests for `getRedactedInput()`?

File: auth/opa/rpcauth/rpcauth.go (hunk 2 above)
Comment (sfc-gh-elinardi) [reply auth/opa/rpcauth/rpcauth.go]: Yup makes sense. Moved it into redact.go. Can't think of better names.

File: auth/opa/rpcauth/rpcauth.go (hunk 1 above)
Comment (sfc-gh-elinardi) [reply auth/opa/rpcauth/rpcauth.go]: Thanks for catching this. Yeah I changed it to use protobuf.Empty and removed the special casing.

**Comment by sfc-gh-elinardi** [comment]
> Looks okay overall. Could you add tests for `getRedactedInput()`?

Added tests for getRedactedInput

File: auth/opa/rpcauth/redact_test.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+   under the License.
+*/
+
+package rpcauth
+
+import (
+	"context"
+	"testing"
+
+	httpPB "github.com/Snowflake-Labs/sansshell/services/httpoverrpc"
```
Comment (sfc-gh-srhodes) [inline auth/opa/rpcauth/redact_test.go:23]: style nit: I prefer all-lowercase for import names

File: auth/opa/rpcauth/rpcauth_test.go [hunk 4]
```diff
@@ -138,6 +139,8 @@ func TestAuthzHook(t *testing.T) {
 		},
 	})
 	testutil.FatalOnErr("json.Marshal extensions", err, t)
+	emptyMessage := emptypb.Empty{}
+	marshaledEmptyMessage, _ := protojson.MarshalOptions{UseProtoNames: true}.Marshal(&emptyMessage)
```
Comment (sfc-gh-srhodes) [inline auth/opa/rpcauth/rpcauth_test.go:143]: The tests might be a bit clearer without this indirection. `[]byte("{}")` should be a valid value.

**Review by sfc-gh-srhodes — APPROVED** [review]

File: auth/opa/rpcauth/rpcauth_test.go (hunk 4 above)
Comment (sfc-gh-elinardi) [reply auth/opa/rpcauth/rpcauth_test.go]: ah I didn't know. thanks!

**Pull request merged**